### PR TITLE
BUG+DEPR: undeprecate item, fix dt64/td64 output type

### LINKED ISSUE #29250: API: consider undeprecating Series.item() ?
I recently ran into this as well (but forgot to open an issue), and raised now by @alexiswl in https://github.com/pandas-dev/pandas/issues/18262#issuecomment-546746982

`Series.item()` was a consequence of (historically) inheriting from np.ndarray, and was deprecated (like a set of other ndarray-inhertited methods/attributes) a while ago.

While `.item()` could also be used to select the "i-th" element (`.item(i)`), and this use case is certainly redundant (not arguing here to get that aspect back), there is one use case where `item()` can actually be useful: if you do not pass *i*, the method returns the element of the Series *only* if it has one element, otherwise it errors.

Such a situation can typically occur if you use boolean indexing (or `query`) to select a single element. Eg in cases like `s[s == 'val']` or `df.loc[df['col1'] == 'val', 'col2']` where you know the condition *should* yield a single element. 
You then typically want the scalar element as result, but those two code snippets give you a Series of one element. In those cases, you could use `item()` to retrieve it: `s[s == 'val'].item()`.

I saw some people using `.item()` exactly for this use case, so wondering if it is worth to keep it for this (only the version without passing *i*).

The logical alternative is doing a `.iloc[0]`, but `.item()` has the advantage of guaranteeing there was actually only one result item.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #29250
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Does the un-deprecation need a dedicated discussion, or is there consensus on that?

I'd be happy to split this into separate bug/depr pieces of reviewers prefer.

### CODE REVIEW

File: pandas/core/base.py [hunk 1]
```diff
@@ -677,19 +676,20 @@ def item(self):
...
+        ValueError
+            If the data is not length-1.
         """
-        warnings.warn(
-            "`item` has been deprecated and will be removed in a future version",
-            FutureWarning,
-            stacklevel=2,
-        )
-        return self.values.item()
+        if len(self) == 1:
```
Comment (WillAyd) [inline pandas/core/base.py]: Was adding this condition discussed somewhere? I would have thought just keep existing behaviour

Comment (jbrockmendel) [reply pandas/core/base.py]: This is the bugfix part.  dt64, dt64tz, and td64 we're currently incorrectly returning int

Comment (WillAyd) [reply pandas/core/base.py]: Hmm doesn't this break non-DTA though? 

```python
>>> type(pd.Series(range(1)).item())
<class 'int'>
>>> type(pd.Series(range(1))[0])
<class 'numpy.int64'>
```

I thought one of the points of item was to return a Python object (at least in the Numpy world)

Comment (WillAyd) [reply pandas/core/base.py]: ^ current behavior

Comment (jorisvandenbossche) [reply pandas/core/base.py]: Yes, I think we should keep the behaviour of `item` to return a python scalar (where possible of course, so for datetime/timedelta it is fine to return a pandas Timestamp/Timedelta I think)

**Comment by jorisvandenbossche** [comment]
This issue is https://github.com/pandas-dev/pandas/issues/29250

File: pandas/core/indexes/period.py [hunk 2]
```diff
@@ -890,23 +889,11 @@ def __setstate__(self, state):
 
     def item(self):
         """
-        Return the first element of the underlying data as a python
-        scalar
-
-        .. deprecated:: 0.25.0
-
+        Return the first element of the underlying data as a python scalar.
```
Comment (shoyer) [inline pandas/core/indexes/period.py]:
```suggestion
        Return the sole element of the underlying data as a python scalar.
```

File: pandas/core/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/base.py]: ok, will update.

File: pandas/core/base.py [hunk 3]
```diff
@@ -677,19 +677,25 @@ def item(self):
...
-        warnings.warn(
-            "`item` has been deprecated and will be removed in a future version",
-            FutureWarning,
-            stacklevel=2,
-        )
-        return self.values.item()
+        if not needs_i8_conversion(self.dtype):
+            # numpy returns ints instead of datetime64/timedelta64 objects,
+            #  which we need to wrap in Timestamp/Timedelta/Period regardless.
+            return self.values.item()
```
Comment (jorisvandenbossche) [inline pandas/core/base.py]: This won't work for ExtensionArrays. We can discuss adding `item` to the interface, but I would rather (or at least for now) let ExtensionArrays take the path you have below that uses iteration (which should already handle the conversion to a python scalar)

Comment (jreback) [reply pandas/core/base.py]: I would be +1 on adding to EA arrays, why have inconsistency in code paths.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/base.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/base.py]: > This won't work for ExtensionArrays

This uses `.values`, so will convert to ndarray and then call `item`.  So it shouldn't be any more broken than what we have now.

Comment (jbrockmendel) [reply pandas/core/base.py]: i have no objection to adding item to EAs separately

Comment (jorisvandenbossche) [reply pandas/core/base.py]: > So it shouldn't be any more broken than what we have now.

And to fix that, the only thing that is needed is adding a `and not is_extension_array_dtype(self.dtype):` to the above `if` check.

Comment (jbrockmendel) [reply pandas/core/base.py]: im happy to do that here, will need tests in a follow-up

File: pandas/core/base.py [hunk 4]
```diff
@@ -677,19 +677,27 @@ def item(self):
...
+            If the data is not length-1.
         """
-        warnings.warn(
-            "`item` has been deprecated and will be removed in a future version",
-            FutureWarning,
-            stacklevel=2,
-        )
-        return self.values.item()
+        if not (
+            is_extension_array_dtype(self.dtype) or needs_i8_conversion(self.dtype)
```
Comment (jreback) [inline pandas/core/base.py]: isn't this redundant? as all needs_i8_conversion are *already* EA

**Review by jreback — COMMENTED** [review]
lgtm, question

**Comment by jreback** [comment]
also needs a rebase

File: pandas/core/base.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/base.py]: no, dt64 and td64 are need i8 conversion but are not EA

File: pandas/core/base.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/base.py]: Has this been resolved?

**Comment by TomAugspurger** [comment]
LGTM, assuming that https://github.com/pandas-dev/pandas/pull/30175/files/0d2867d5518ac8e5032e7dfca5d0adb5811fa84d#diff-9c718a39eb63c1bfac4fbeacf2906ffd 
has been addressed.

File: pandas/core/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/base.py]: I think the concerns raised by @WillAyd and @jorisvandenbossche have been addressed

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks @jbrockmendel !

(will look into adding a EA test)

**Pull request merged**